src/main.rs

```diff
@@ -1,20 +1,51 @@
-//#[feature(conservative_impl_trait, generators, generator_trait)]
+#![feature(conservative_impl_trait, generators, generator_trait)]
+
+use std::ops::Generator;
 
 fn main() {
     println!("Hello, world!");
     let funct = |t: f64, x: f64| -> f64 { t*x };
-
-    let mut x_this = 1.0;
-    let mut t_this = 0.0;
+    let mut x_gen = Euler{}.integrate_to(&funct, 1.1, 0.0, 5.06, 0.1);
     for _ in 0..10 {
-        let (x_next, t_next) = euler_step(&funct, x_this, 0.1, t_this);
-        println!("{} {}", x_next, t_next);
-        x_this = x_next;
-        t_this = t_next;
+        println!("{:?}", x_gen);
     }
 }
 
+trait Integrator {
+    fn step<F: Fn(f64, f64) -> f64>(&self, f: &F, x0: f64, t0: f64, dt: f64) -> f64;
 
-fn euler_step<F: Fn(f64, f64) -> f64>(f: &F, x0: f64, dt: f64, t: f64) -> (f64, f64) {
-    (x0 + dt * f(t, x0), t + dt)
+    fn integrate<'a, F: Fn(f64, f64) -> f64>(&'a self, f: &'a F, x0: f64, t0: f64, dt: f64) -> Box<Generator<Return = f64, Yield = f64> + 'a>;
+
+    fn integrate_to<F: Fn(f64, f64) -> f64>(&self, f: &F, x0: f64, t0: f64, t_end: f64, dt: f64) -> f64;
+}
+
+
+struct Euler {
+}
+
+impl Integrator for Euler {
+    fn step<F: Fn(f64, f64) -> f64>(&self, f: &F, x0: f64, t0: f64, dt: f64) -> f64 {
+        x0 + dt * f(t0, x0)
+    }
+
+    fn integrate<'a, F: Fn(f64, f64) -> f64>(&'a self, f: &'a F, mut x0: f64, mut t0: f64, dt: f64) -> Box<Generator<Return = f64, Yield = f64> + 'a> {
```

Comment:
You could implement this and `integrate_to` as default implementations into `Integrator`, since this will probably be the same for every implementation?